pulsar-client/src/main/java/org/apache/pulsar/client/impl/MessageImpl.java

```diff
@@ -269,31 +269,26 @@ public MessageImpl(String topic, String msgId, Map<String, String> properties,
         return msg;
     }
 
-    public static MessageImpl<byte[]> deserializeBrokerEntryMetaDataFirst(
-            ByteBuf headersAndPayloadWithBrokerEntryMetadata) throws IOException {
-        @SuppressWarnings("unchecked")
-        MessageImpl<byte[]> msg = (MessageImpl<byte[]>) RECYCLER.get();
-
-        msg.brokerEntryMetadata =
+    public static long getEntryTimestamp( ByteBuf headersAndPayloadWithBrokerEntryMetadata) throws IOException {
+        // get broker timestamp first if BrokerEntryMetadata is enabled with AppendBrokerTimestampMetadataInterceptor
+        BrokerEntryMetadata brokerEntryMetadata =
                 Commands.parseBrokerEntryMetadataIfExist(headersAndPayloadWithBrokerEntryMetadata);
-
-        if (msg.brokerEntryMetadata != null) {
-            msg.msgMetadata.clear();
-            msg.payload = null;
-            msg.messageId = null;
-            msg.topic = null;
-            msg.cnx = null;
-            msg.properties = Collections.emptyMap();
-            return msg;
+        if (brokerEntryMetadata != null && brokerEntryMetadata.hasBrokerTimestamp()) {
+            return brokerEntryMetadata.getBrokerTimestamp();
         }
-
+        // otherwise get the publish_time
+        MessageImpl<byte[]> msg = (MessageImpl<byte[]>) RECYCLER.get();
```

Comment:
There's a memory leak here.

Before the change, the `MessageImpl<byte[]>` returned by `deserializeBrokerEntryMetaDataFirst` before will call `recycle()` method in caller side like `expireMessages`, `findMessages`, etc. 

After the change, `msg` here never calls `recycle()` method.

Reply:
You're right!  Fixed.

Reply:
Why not just init a `MessageMetadata` object instead of `MessageImpl` ?